Gamepad: Align closer to spec and implement missing slots

msub2 opened the pull request:
<!-- Please describe your changes on the following line: -->
This follows up from the initial Gamepad implementation and addresses some TODOs around `gamepad.[[exposed]]` and `navigator.[[hasGamepadGesture]]`, as well as some missing spec steps. It also properly gates gamepad handling behind the `dom.gamepad.enabled` pref, as previously the `gamepadconnected` and `gamepaddisconnected` events would fire regardless.

This also allows one more WPT test to pass, `gamepad/not-fully-active`.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

msub2 committed:
Implement missing gamepad slots, align to spec more

- Fixes TODO's from initial gamepad implementation
- Adds some missing spec steps

msub2 committed:
Only handle gamepad events when pref is enabled

msub2 committed:
Return empty list in getGamepads if document not active

msub2 committed:
./mach fmt

Taym95 commented on components/script/dom/globalscope.rs:
```diff
@@ -3142,15 +3143,16 @@ impl GlobalScope {
-
                     if window.Document().is_fully_active() {
-                        gamepad.update_connected(true);
+                        gamepad.update_connected(true, has_gesture);
```
here is `has_gesture ` and when you call function you use `should_notify ` I would expect prams name to be the same

gterzian replied:
agree, arg name should be `has_gesture`.

Taym95 commented on components/script/dom/globalscope.rs:
```diff
@@ -3211,8 +3214,26 @@ impl GlobalScope {
+                                                }),
+                                                window.upcast(),
+                                            )
+                                            .unwrap();
```
maybe you should check if gamepad_task_source is not `None`

gterzian replied:
This is about the queuing of the task, but we should expect with a clear message.

Taym95 commented on components/script/dom/navigator.rs:
```diff
@@ -171,11 +190,16 @@ impl NavigatorMethods for Navigator {
-        let root = self
-            .gamepads
-            .or_init(|| GamepadList::new(&self.global(), &[]));
+        let global = self.global();
```
I would expect some wpt tests for these changes

msub2 replied:
The existing WPT that covers this fails on an assert that getGamepads() returns an array. Currently it returns a GamepadList, and attempting to change it to a frozen array of gamepads is non-trivial due to `to_frozen_array` seemingly being incompatible with the Gamepad object.

msub2 replied:
This is now covered by the aforementioned WPT test (`gamepad/not_fully_active.https.html`)

Taym95 commented on components/script/dom/gamepad.rs:
```diff
@@ -265,4 +269,12 @@ impl Gamepad {
         }
     }
+
+    pub fn exposed(&self) -> bool {
```
I think we should add doc link when we add public API, not link for `exposed ` ?

msub2 replied:
Added

msub2 committed:
Update getGamepads to return an array instead of GamepadList

msub2 committed:
Add spec link for [[exposed]] slot

msub2 committed:
Remove failing test expectations for not-fully-active

gterzian commented on components/script/dom/navigator.rs:
```diff
@@ -63,6 +67,7 @@ impl Navigator {
             permissions: Default::default(),
             mediasession: Default::default(),
             gpu: Default::default(),
+            has_gamepad_gesture: Cell::new(false),
```
/// <https://w3c.github.io/gamepad/#dfn-hasgamepadgesture>

gterzian commented on components/script/dom/navigator.rs:
```diff
@@ -47,6 +50,7 @@ pub struct Navigator {
     permissions: MutNullableDom<Permissions>,
```
Above /// <https://w3c.github.io/gamepad/#dfn-gamepads>

gterzian commented on components/script/dom/navigator.rs:
```diff
@@ -170,13 +190,18 @@ impl NavigatorMethods for Navigator {
-        let root = self
-            .gamepads
-            .or_init(|| GamepadList::new(&self.global(), &[]));
+    fn GetGamepads(&self) -> Vec<Option<DomRoot<Gamepad>>> {
```
https://w3c.github.io/gamepad/#dom-navigator-getgamepads

gterzian commented on components/script/dom/navigator.rs:
```diff
@@ -170,13 +190,18 @@ impl NavigatorMethods for Navigator {
+        if !doc.is_fully_active() || !self.has_gamepad_gesture.get() {
+            return Vec::new();
+        }
+
```
TODO for permissions

gterzian commented on components/script/dom/gamepadlist.rs:
```diff
@@ -48,6 +48,15 @@ impl GamepadList {
+    pub fn list(&self) -> Vec<Option<DomRoot<Gamepad>>> {
+        self.list
+            .borrow()
+            .clone()
```
No need to clone since you can build the list from references.

gterzian commented on components/script/dom/gamepadlist.rs:
```diff
@@ -48,6 +48,15 @@ impl GamepadList {
         self.list.borrow_mut().remove(index);
     }
+
+    pub fn list(&self) -> Vec<Option<DomRoot<Gamepad>>> {
```
I would say that in a follow-up we can remove gamepadlist, since a `RefCell<Vec<Dom<Gamepad>>>` on the navigator would seem to do the trick.

msub2 replied:
Agreed, I'll probably do that after this and before GamepadHapticActuator

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -3142,15 +3142,16 @@ impl GlobalScope {
-                    // TODO: 3.4 If navigator.[[hasGamepadGesture]] is true:
-                    // TODO: 3.4.1 Set gamepad.[[exposed]] to true.
-
                     if window.Document().is_fully_active() {
```
this should be inside the  `has_gesture` block.

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -3211,8 +3213,26 @@ impl GlobalScope {
-                            // TODO: 6. If navigator.[[hasGamepadGesture]] is false
-                            //          and gamepad contains a gamepad user gesture:
+                            if !window.Navigator().has_gamepad_gesture() {
+                                window.Navigator().set_has_gamepad_gesture(true);
```
We should only set it to true according to https://w3c.github.io/gamepad/#dfn-gamepad-user-gesture. Is this dealt with at the embedding layer? It might relate to what you experienced initially as the pad not responding to a first input.

msub2 replied:
I suppose it's a bit of a simplification as it is now, I'm treating any input as a gamepad user gesture but I should technically be making it more rigorous and checking for axis threshold too. This is unrelated to my initial issue with it not responding to first input though, that was caused by GilRs not emitting a changed event when a button doesn't have prior state (e.g. pressed for the very first time).

gterzian reviewed (changes requested):
Great start! Couple of comments...

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -3211,8 +3213,26 @@ impl GlobalScope {
+                                                }),
+                                                window.upcast(),
+                                            )
+                                            .unwrap();
```
Let's also use `queue_with_canceller` for all tasks(here and others), and replace `unwrap` with an appropriate `expect`.

msub2 committed:
A few fixes

- Change should_notify to has_gesture
- Add spec links and TODO to navigator
- Remove unneeded clone from GamepadList::list
- Move gamepadconnected event firing into has_gesture block

msub2 committed:
Use queue_with_canceller for tasks and add expects

msub2 committed:
Explicitly check for gamepad user gesture

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -3193,32 +3194,54 @@ impl GlobalScope {
                             let current_time = global.performance().Now();
                             gamepad.update_timestamp(*current_time);
 
+                            let mut contains_user_gesture = false;
```
let's put this in a separate function taking the update type as argument and with a doclink to https://w3c.github.io/gamepad/#dfn-gamepad-user-gesture

gterzian reviewed (approved):
LGTM with one nit.

msub2 committed:
Move user gesture check into separate function

gterzian commented on components/script/dom/globalscope.rs:
```diff
@@ -3244,6 +3240,18 @@ impl GlobalScope {
     }
 
+    /// <https://www.w3.org/TR/gamepad/#dfn-gamepad-user-gesture>
+    pub fn contains_user_gesture(&self, update_type: GamepadUpdateType) -> bool {
```
Is clippy not complaining about not using `self`? 

I think this should be a function(not method), inside `gamepad`. Because now it reads as if it's the global that "contains a user gesture" or not. 

also that way you only need to import that function, and the constants are used only in `gamepad`.

gterzian reviewed (changes requested):
Ok thanks for the refactor, just one last thing and then good to go!

msub2 committed:
Change contains_user_gesture to be a gamepad function

msub2 committed:
mach fmt

gterzian commented on components/script/dom/gamepad.rs:
```diff
@@ -263,4 +271,26 @@ impl Gamepad {
+            return value > BUTTON_PRESS_THRESHOLD;
+        },
+    };
 }
```
Let's just add a newline here.

msub2 replied:
A second one? It ends at 297

gterzian replied:
Ah ok, should be fine then...

gterzian reviewed (approved):
LGTM, one nit.

gterzian commented on components/script/dom/gamepad.rs:
```diff
@@ -23,9 +24,12 @@ use crate::dom::gamepadpose::GamepadPose;
 
+// This value is for determining when to consider a gamepad as having a user gesture
+// from an axis tilt. This matches the threshold in Chromium.
+pub const AXIS_TILT_THRESHOLD: f64 = 0.5;
```
One more thing: this can be private to the module.

msub2 committed:
Change axis/button threshold constants to be private to module

gterzian reviewed (approved):
Thank you!
